torch/custom_class_detail.h

```diff
@@ -0,0 +1,72 @@
+#pragma once
+
+#include <c10/util/Metaprogramming.h>
+#include <c10/util/TypeTraits.h>
+
+namespace torch {
+namespace jit {
+
+namespace detail {
+
+// Argument type utilities
+template <class R, class...>
+struct types {
+  using type = types;
+};
+
+template <class Sig>
+struct args;
+
+// Method
+template <class R, class CurClass, class... Args>
+struct args<R (CurClass::*)(Args...)> : types<R, Args...> {};
+
+// Const method
+template <class R, class CurClass, class... Args>
+struct args<R (CurClass::*)(Args...) const> : types<R, Args...> {};
+
+template <class Sig>
+using args_t = typename args<Sig>::type;
+
+template <
+    typename Get,
+    typename Set,
+    typename = typename c10::guts::infer_function_traits<Get>::type::func_type,
+    typename = typename c10::guts::infer_function_traits<Set>::type::func_type>
+struct pickle_factory;
+
+template <
+    typename Get,
+    typename Set,
+    typename RetState,
+    typename Self,
+    typename NewInstance,
+    typename ArgState>
+struct pickle_factory<Get, Set, RetState(Self), NewInstance(ArgState)> {
+  pickle_factory(Get&& g, Set&& s) : g(std::move(g)), s(std::move(s)) {}
+
+  [[noreturn]] ArgState arg_state_type() {}
```

Comment:
nit: why do you need `decltype(arg_state_type())` instead of just putting here `using arg_state = ArgState` and doing `decltype(pickle_factor).arg_state`?